idd/Energy+.idd.in

```diff
@@ -107737,3 +107737,218 @@ Output:PreprocessorMessage,
         \retaincase
    A12; \field Message Line 10
         \retaincase
+
+\group Python Plugin System
+
+PythonPlugin:SearchPaths,
```

Comment:
I  understand why there's a desire to have this as a capability, but I am concerned that this will lead to some fairly difficult support situations - e.g. having to replicate a user's file system to run an input file. It'll get worse if file operations are happening with relative paths. What would be the downside of having a fixed directory (on a per user basis) for plugins? Just one directory won't eliminate issues, but would limit the scope significantly.

Reply:
@jasondegraw I would expect most every user to simply rely on the defaults -- the search path will include the directory of the energyplus binary (necessary for this to work) as well as the directory of the given input file.  I anticipate every user who tries this out will want to just keep their python stuff right next to the IDF.  But a major enhancement here is enabling  power users to keep libraries of component models and functionality to be reused in multiple simulations.  In this case, I wouldn't want to tell them they need to continually make more copies of their python code just to drop in folder after folder.  If we just handle paths properly, why would we want to limit the possible use cases, anyway?  I don't share your concerns about paths, and I'll be the one doing support on files related to the plugins.

Reply:
@Myoldmopar I'm not advocating for requiring multiple copies or suggesting that paths wouldn't be handled properly. Quite the opposite, I'm assuming that paths would be handled correctly. Having one "plugin directory" per user wouldn't require copying of files, it just limits where things can be. It'd still have to be "settable" somehow, it just wouldn't be on a per-model basis. And no, you wouldn't be the only one doing support.

Reply:
Are you thinking like $HOME/EnergyPlusPlugins?  I know tools have tended to do that, and I always ended up with random **measure** or **bcl** folders, right in my personal home directory, that I otherwise kept pristine, but I digress...  

I just don't want to force the user into either of these situations: 
 - having to put _every_ python script they do for _all_ versions of E+ into _one_ single folder on their operating system, or
 - having to keep the python exactly right next to every IDF they run

Even if we went down the path of the $HOME/EnergyPlusPlugins path, the user would need to submit their python scripts for support, so they'd have to go collect it and submit it.  If they've created a nice Python package, and they zip that, along with the IDF up for us, and we extract it together into one directory, along with a weather file, then we can run it directly without adding any search paths.  If they used a custom one in their original IDF, we can just delete that search paths object from their IDF and debug directly.  This is similar to Mike removing schedule:file paths, but actually easier. I think it'll be great.

Reply:
Yes, something like that, but it'd probably need to be versioned, and an absolute no on UUIDs as directory names. I don't see a problem in having one plugin directory + the IDF directory as the locations to look in, I don't recall objecting to that.

Adding this feature (pretty much in any form that I can think of) allows for the creation of models that are very tightly tied to the filesystem on which they are run, I'm just advocating for setting tighter limits to keep that dependency more tractable. If more complexity is allowed, the models will get more complex. It's only easier to deal with if the Python code can be moved, which is an assumption I don't think can be made. I'd hope that we'd advocate for that (i.e. code that isn't tied to the filesystem) as a good programming practice, but can you enforce that while still giving full access to the Python standard library?

In any case, OK, yeah, I get it, you're building a house.